### PR TITLE
change default of shuffle to False and the description

### LINKED ISSUE #31232: Default shuffle behavior of DistributedSampler
## 🚀 Feature
The **DistributedSampler** should shuffle data without explicitly set `set_epoch`.

## Motivation
I have found that the parameter shuffle of **DistributedSampler** is set True in default but the behavior and intuition are inconsistent. The GPU always load the same data between every epoch. It still needs manual set sampler.set_epoch to change the data loaded by GPU
## Pitch
Truely shuffle the data when the shuffle=True in **DistributedSampler** 
## Alternatives
Just show this potential feature in doc.
## Additional context

The test code, 

`CUDA_VISIBLE_DEVICES=1,3 python -m torch.distributed.launch --nproc_per_node=2 test.py`

test.py:
```import torch
import torch.nn as nn
from torch.utils.data import Dataset, DataLoader
from torch.utils.data.distributed import DistributedSampler


torch.distributed.init_process_group(backend="nccl")

input_size = 5
output_size = 2
batch_size = 2
data_size = 16

local_rank = torch.distributed.get_rank()
torch.cuda.set_device(local_rank)
device = torch.device("cuda", local_rank)

class RandomDataset(Dataset):
    def __init__(self, size, length, local_rank):
        self.len = length
        self.data = torch.stack([torch.ones(5), torch.ones(5)*2,
                                 torch.ones(5)*3,torch.ones(5)*4,
                                 torch.ones(5)*5,torch.ones(5)*6,
                                 torch.ones(5)*7,torch.ones(5)*8,
                                 torch.ones(5)*9, torch.ones(5)*10,
                                 torch.ones(5)*11,torch.ones(5)*12,
                                 torch.ones(5)*13,torch.ones(5)*14,
                                 torch.ones(5)*15,torch.ones(5)*16]).to('cuda')

        self.local_rank = local_rank
    def __getitem__(self, index):

        return self.data[index]

    def __len__(self):
        return self.len
    
dataset = RandomDataset(input_size, data_size, local_rank)
sampler = DistributedSampler(dataset)
rand_loader = DataLoader(dataset=dataset,
                         batch_size=batch_size,
                         sampler=sampler)

e = 0
while e < 2:
    t = 0
    sampler.set_epoch(e)
    for data in rand_loader:
        print(data)
    e+=1
```

### PR DESCRIPTION
**Pull request by ydaiming**
Fix https://github.com/pytorch/pytorch/issues/31232#

A small change in the default is made. Wait for auto checks.

@SsnL

### CODE REVIEW

File: torch/utils/data/distributed.py [hunk 1]
```diff
@@ -47,7 +47,7 @@ class DistributedSampler(Sampler):
         ...     train(loader)
     """
 
-    def __init__(self, dataset, num_replicas=None, rank=None, shuffle=True, seed=0):
+    def __init__(self, dataset, num_replicas=None, rank=None, shuffle=False, seed=0):
```
Comment (rohan-varma) [inline torch/utils/data/distributed.py:50]: Does the default change present a backwards compatibility concern? This means previously written code like: 
```
sampler = DistributedSampler()
for epoch in epochs:
    sampler.set_epoch(epoch)
```

won't work as expected since shuffle would be false now.

Comment (ydaiming) [reply torch/utils/data/distributed.py:50]: This is a very reasonable concern. Maybe we can change the default `seed` to `1`, so the data is truly shuffled on epoch 0.

**Comment by rohan-varma** [comment]
@mrshenli @zhaojuanmao Could you help take a look at this PR? Thanks!

File: torch/utils/data/distributed.py [hunk 2]
```diff
@@ -47,7 +47,7 @@ class DistributedSampler(Sampler):
         ...     train(loader)
     """
 
-    def __init__(self, dataset, num_replicas=None, rank=None, shuffle=True, seed=0):
+    def __init__(self, dataset, num_replicas=None, rank=None, shuffle=True, seed=1):
```
Comment (rohan-varma) [inline torch/utils/data/distributed.py:50]: Why does changing the random seed from 0 to 1 change any behavior here? `g.manual_seed(0)` is still valid for reproducibility purposes, no?

**Comment by mrshenli** [comment]
I might miss sth, looks like this PR only changes the random seed. Why this would fix #31232?

**Review by mrshenli — COMMENTED** [review]
If the goal is to fix #31232, we will also need a test case.

**Comment by mrshenli** [comment]
also adding @vincentqb to review for dataloader PRs

**Comment by vincentqb** [comment]
> I might miss sth, looks like this PR only changes the random seed. Why this would fix #31232?

I second this. Maybe adding a test could prove whether this indeed fixes the issue?
